File: AlumniRatio.Backend/AlumniRatio.Backend/Journals/General/TransactionAnalysis/OfficeSuppliesPurchaseEntryAnalyzer.cs
using AlumniRatio.Backend.Accounts;
using AlumniRatio.Backend.Transactions;

namespace AlumniRatio.Backend.Journals.General.TransactionAnalysis;

public class OfficeSuppliesPurchaseEntryAnalyzer : IGeneralJourneyEntryAnalyzer
{
    public GeneralJournalEntry CreateEntryBasedOnTransaction(int id, Transaction transaction)
    {
        var journalEntry = new GeneralJournalEntry(id);
        
        if (transaction.PaymentType.GetValueOrDefault().Equals(PaymentType.Debit))
        {
            var officeSuppliesAccount = new GeneralJournalRow(AccountBook.OFFICE_SUPPLIES_ACCOUNT.Number, transaction.Debit.GetValueOrDefault(), 0, AccountBook.OFFICE_SUPPLIES_ACCOUNT.Name);
            var cashEntry = new GeneralJournalRow(AccountBook.CASH_ACCOUNT.Number, 0 , transaction.Debit.GetValueOrDefault(), AccountBook.CASH_ACCOUNT.Name);
            journalEntry.Explanation = "Para registrar compra de materiales a débito.";
            journalEntry.AddRow(officeSuppliesAccount);
            journalEntry.AddRow(cashEntry);
        }
        
        else if (transaction.PaymentType.GetValueOrDefault().Equals(PaymentType.Credit))
        {
            var officeSuppliesAccount = new GeneralJournalRow(AccountBook.OFFICE_SUPPLIES_ACCOUNT.Number, transaction.Credit.GetValueOrDefault(), 0, AccountBook.OFFICE_SUPPLIES_ACCOUNT.Name);
            var payableAccount = new GeneralJournalRow(AccountBook.PAYABLE_ACCOUNT.Number, 0 , transaction.Credit.GetValueOrDefault(), AccountBook.PAYABLE_ACCOUNT.Name);
            journalEntry.Explanation = "Para registrar compra de materiales a crédito.";
            journalEntry.AddRow(officeSuppliesAccount);
            journalEntry.AddRow(payableAccount);
        }

        else
        {
            var total = (transaction.Debit.GetValueOrDefault() + transaction.Credit.GetValueOrDefault());
            var officeSuppliesAccount = new GeneralJournalRow(AccountBook.OFFICE_SUPPLIES_ACCOUNT.Number, total, 0, AccountBook.OFFICE_SUPPLIES_ACCOUNT.Name);
            var cashEntry = new GeneralJournalRow(AccountBook.CASH_ACCOUNT.Number, 0 , transaction.Debit.GetValueOrDefault(), AccountBook.CASH_ACCOUNT.Name);
            var payableAccount = new GeneralJournalRow(AccountBook.PAYABLE_ACCOUNT.Number, 0 , transaction.Credit.GetValueOrDefault(), AccountBook.PAYABLE_ACCOUNT.Name);
            journalEntry.Explanation = "Para registrar compra de materiales a crédito y débito.";
            journalEntry.AddRow(officeSuppliesAccount);
            journalEntry.AddRow(cashEntry);
            journalEntry.AddRow(payableAccount);
        }
        return journalEntry;
    }
}

File: alumni-ratio-backend/src/main/java/edu/pucmm/eict/alumniratio/journals/general/analyzer/OfficeSuppliesPurchaseEntryAnalyzer.java
package edu.pucmm.eict.alumniratio.journals.general.analyzer;

import edu.pucmm.eict.alumniratio.accounts.AccountBook;
import edu.pucmm.eict.alumniratio.exercises.Exercise;
import edu.pucmm.eict.alumniratio.journals.general.GeneralJournalEntry;
import edu.pucmm.eict.alumniratio.journals.general.GeneralJournalRow;
import edu.pucmm.eict.alumniratio.transactions.Transaction;

import java.math.BigDecimal;
import java.time.LocalDate;

public class OfficeSuppliesPurchaseEntryAnalyzer implements PaymentTypeEntryAnalyzer {

    @Override
    public GeneralJournalEntry debit(Transaction transaction, Exercise exercise) {
        var entryExplanation = "Compra de materiales de oficina al contado";
        var transactionDate = LocalDate.of(exercise.getYearOperations(), exercise.getMonthOperations(), transaction.getDay());
        var entry = new GeneralJournalEntry(transactionDate, entryExplanation);
        var cashAccountName = "Efectivo";
        var officeSuppliesAccountName = "Materiales de oficina";
        entry.addRow(new GeneralJournalRow(transactionDate, AccountBook.OFFICE_SUPPLIES, officeSuppliesAccountName , transaction.getDebit(), BigDecimal.ZERO));
        entry.addRow(new GeneralJournalRow(transactionDate, AccountBook.CASH, cashAccountName, BigDecimal.ZERO, transaction.getDebit()));
        return entry;
    }

    @Override
    public GeneralJournalEntry credit(Transaction transaction, Exercise exercise) {
        var entryExplanation = "Compra de materiales de oficina a crédito";
        var transactionDate = LocalDate.of(exercise.getYearOperations(), exercise.getMonthOperations(), transaction.getDay());
        var entry = new GeneralJournalEntry(transactionDate, entryExplanation);
        var accountPayablesName = "Cuentas por pagar";
        var officeSuppliesAccountName = "Materiales de oficina";
        entry.addRow(new GeneralJournalRow(transactionDate, AccountBook.OFFICE_SUPPLIES, officeSuppliesAccountName , transaction.getCredit(), BigDecimal.ZERO));
        entry.addRow(new GeneralJournalRow(transactionDate, AccountBook.ACCOUNTS_PAYABLES, accountPayablesName, BigDecimal.ZERO, transaction.getCredit()));
        return entry;
    }

    @Override
    public GeneralJournalEntry mixed(Transaction transaction, Exercise exercise) {
        var entryExplanation = "Compra de materiales de oficina al contado y crédito.";
        var transactionDate = LocalDate.of(exercise.getYearOperations(), exercise.getMonthOperations(), transaction.getDay());
        var entry = new GeneralJournalEntry(transactionDate, entryExplanation);
        var cashAccountName = "Efectivo";
        var officeSuppliesAccountName = "Materiales de oficina";
        var accountPayablesName = "Cuentas por pagar";
        var amount = transaction.getCredit().add(transaction.getDebit());
        entry.addRow(new GeneralJournalRow(transactionDate, AccountBook.OFFICE_SUPPLIES, officeSuppliesAccountName , amount, BigDecimal.ZERO));
        entry.addRow(new GeneralJournalRow(transactionDate, AccountBook.ACCOUNTS_PAYABLES, accountPayablesName, BigDecimal.ZERO, transaction.getCredit()));
        entry.addRow(new GeneralJournalRow(transactionDate, AccountBook.CASH, cashAccountName, BigDecimal.ZERO, transaction.getDebit()));
        return entry;
    }
}
